tutorial/tests/testsuite.py

```diff
@@ -49,9 +67,136 @@ def pytest_generate_tests(self, metafunc: pytest.Metafunc) -> None:
             metafunc.parametrize("function_to_test", [self.function_to_test])
 
 
+@dataclasses.dataclass
+class TestResult:
+    """Container class to store the test results when we collect them"""
+    stdout: str
+    stderr: str
+    test_name: str
+    success: bool
+
+
+@dataclasses.dataclass
+class OutputConfig:
+    """Container class to store the information to display in the test output"""
+    style: str
+    name: str
+    result: str
+
+
+def format_success_failure(syntax_error: bool, success: bool, name: str) -> OutputConfig:
+    """
+    Depending on the test results, returns a fragment that represents
+    either an error message, a success message, or a syntax error warning
+    """
+    if syntax_error:
+        return OutputConfig(
+            "alert-warning",
+            f"Tests <strong>CANNOT RUN</strong> for this cell.",
+            "&#129300 Careful, looks like you have a syntax error.",
+        )
+
+    if success:
+        return OutputConfig(
+            "alert-success",
+            f"Tests <strong>PASSED</strong> for the function <code>{name}</code>",
+            "&#x1F64C Congratulations, your solution was correct!",
+        )
+    else:
+        return OutputConfig(
+            "alert-danger",
+            f"Tests <strong>FAILED</strong> for the function <code>{name}</code>",
+            "&#x1F631 Your solution was not correct!",
+        )
+
+
+def format_long_stdout(text: str) -> str:
+    """
+    Format a long test stdout as a HTML by using the <details> element
+    """
+
+    def join_lines(text: str) -> str:
+        text = text.splitlines()
+        return "".join(
+            [f"<p>{line}</p>" for line in text[-6:-1]]
+        )
+
+    separator = 59 * ' _'
```

Comment:
Why the fixed value of 59? Is it because the test output is separated by 59 `-`? Can we make it more flexible?

Reply:
This was the only unique separator I could find, so that we can keep only the useful output. Everything before the '_ _ _ ...' is not needed and gets ignored. Not an optimal solution, I know, but I can work on it if you'd prefer something more flexible.

Reply:
It's fine like this for the time being, thank you!